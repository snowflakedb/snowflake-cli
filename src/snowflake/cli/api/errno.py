# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# General errors
NO_WAREHOUSE_SELECTED_IN_SESSION = 606
EMPTY_SQL_STATEMENT = 900

SQL_COMPILATION_ERROR = 1003
OBJECT_ALREADY_EXISTS_IN_DOMAIN = 1998
OBJECT_ALREADY_EXISTS = 2002
DOES_NOT_EXIST_OR_NOT_AUTHORIZED = 2003  # BASE_TABLE_OR_VIEW_NOT_FOUND
DUPLICATE_COLUMN_NAME = 2025
VIEW_EXPANSION_FAILED = 2037
DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED = (
    2043  # OBJECT_DOES_NOT_EXIST_OR_CANNOT_PERFORM_OPERATION
)
INSUFFICIENT_PRIVILEGES = 3001  # NOT_AUTHORIZED
INVALID_OBJECT_TYPE_FOR_SPECIFIED_PRIVILEGE = 3008
ROLE_NOT_ASSIGNED = 3013
NO_INDIVIDUAL_PRIVS = 3028
OBJECT_ALREADY_EXISTS_NO_PRIVILEGES = 3041

# Native Apps
APPLICATION_PACKAGE_MANIFEST_SPECIFIED_FILE_NOT_FOUND = 93003
APPLICATION_FILE_NOT_FOUND_ON_STAGE = 93009
CANNOT_GRANT_OBJECT_NOT_IN_APP_PACKAGE = 93011
CANNOT_GRANT_RESTRICTED_PRIVILEGE_TO_APP_PACKAGE_SHARE = 93012
APPLICATION_PACKAGE_VERSION_ALREADY_EXISTS = 93030
APPLICATION_PACKAGE_VERSION_NAME_TOO_LONG = 93035
APPLICATION_PACKAGE_PATCH_DOES_NOT_EXIST = 93036
APPLICATION_PACKAGE_MAX_VERSIONS_HIT = 93037
CANNOT_UPGRADE_FROM_LOOSE_FILES_TO_VERSION = 93044
CANNOT_UPGRADE_FROM_VERSION_TO_LOOSE_FILES = 93045
ONLY_SUPPORTED_ON_DEV_MODE_APPLICATIONS = 93046
NO_VERSIONS_AVAILABLE_FOR_ACCOUNT = 93054
NOT_SUPPORTED_ON_DEV_MODE_APPLICATIONS = 93055
APPLICATION_NO_LONGER_AVAILABLE = 93079
APPLICATION_INSTANCE_FAILED_TO_RUN_SETUP_SCRIPT = 93082
APPLICATION_INSTANCE_NO_ACTIVE_WAREHOUSE_FOR_CREATE_OR_UPGRADE = 93083
APPLICATION_INSTANCE_EMPTY_SETUP_SCRIPT = 93084
VERSION_REFERENCED_BY_RELEASE_DIRECTIVE = 93088
APPLICATION_PACKAGE_MANIFEST_CONTAINER_IMAGE_URL_BAD_VALUE = 93148
CANNOT_GRANT_NON_MANIFEST_PRIVILEGE = 93118
APPLICATION_OWNS_EXTERNAL_OBJECTS = 93128
APPLICATION_PACKAGE_PATCH_ALREADY_EXISTS = 93168
APPLICATION_PACKAGE_CANNOT_SET_EXTERNAL_DISTRIBUTION_WITH_SPCS = 93197
NATIVE_APPLICATION_MANIFEST_UNRECOGNIZED_FIELD = 93301
NATIVE_APPLICATION_MANIFEST_UNEXPECTED_VALUE_FOR_PROPERTY = 93302
NATIVE_APPLICATION_MANIFEST_GENERIC_JSON_ERROR = 93303
NATIVE_APPLICATION_MANIFEST_INVALID_SYNTAX = 93300
APPLICATION_REQUIRES_TELEMETRY_SHARING = 93321
CANNOT_DISABLE_MANDATORY_TELEMETRY = 93329
VERSION_NOT_ADDED_TO_RELEASE_CHANNEL = 512008
CANNOT_DISABLE_RELEASE_CHANNELS = 512001
RELEASE_DIRECTIVES_VERSION_PATCH_NOT_FOUND = 93036
RELEASE_DIRECTIVE_DOES_NOT_EXIST = 93090
VERSION_DOES_NOT_EXIST = 93031
VERSION_NOT_IN_RELEASE_CHANNEL = 512010
ACCOUNT_DOES_NOT_EXIST = 1999
ACCOUNT_HAS_TOO_MANY_QUALIFIERS = 906
CANNOT_MODIFY_RELEASE_CHANNEL_ACCOUNTS = 512017
VERSION_ALREADY_ADDED_TO_RELEASE_CHANNEL = 512005
MAX_VERSIONS_IN_RELEASE_CHANNEL_REACHED = 512004
MAX_UNBOUND_VERSIONS_REACHED = 512023
CANNOT_DEREGISTER_VERSION_ASSOCIATED_WITH_CHANNEL = 512021
TARGET_ACCOUNT_USED_BY_OTHER_RELEASE_DIRECTIVE = 93091


ERR_JAVASCRIPT_EXECUTION = 100132

SNOWSERVICES_IMAGE_REPOSITORY_IMAGE_IMPORT_TO_NATIVE_APP_FAIL = 397007
SNOWSERVICES_IMAGE_MANIFEST_NOT_FOUND = 397012
SNOWSERVICES_IMAGE_REPOSITORY_FAILS_TO_RETRIEVE_IMAGE_HASH_NEW = 397013

NO_REFERENCE_SET_FOR_DEFINITION = 505019
NO_ACTIVE_REF_DEFINITION_WITH_REF_NAME_IN_APPLICATION = 505026
