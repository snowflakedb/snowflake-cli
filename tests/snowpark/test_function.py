# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from pathlib import Path
from textwrap import dedent
from unittest import mock

import pytest
from snowflake.cli._plugins.snowpark.package_utils import (
    DownloadUnavailablePackagesResult,
)
from snowflake.cli.api.errno import DOES_NOT_EXIST_OR_NOT_AUTHORIZED
from snowflake.connector import ProgrammingError

from tests_common import IS_WINDOWS

if IS_WINDOWS:
    pytest.skip("Requires further refactor to work on Windows", allow_module_level=True)


mock_session_has_warehouse = mock.patch(
    "snowflake.cli.api.sql_execution.SqlExecutionMixin.session_has_warehouse",
    lambda _: True,
)


@pytest.mark.parametrize(
    "project_name", ["snowpark_functions", "snowpark_functions_v2"]
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function(
    mock_object_manager,
    mock_connector,
    mock_ctx,
    runner,
    project_directory,
    project_name,
):
    mock_object_manager.return_value.describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    with project_directory(project_name) as project_dir:
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
            catch_exceptions=False,
        )

    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string default 'default value', b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=()
            """
        ).strip(),
    ]


@pytest.mark.parametrize(
    "project_name",
    ["snowpark_function_external_access", "snowpark_function_external_access_v2"],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function_with_external_access(
    mock_object_manager,
    mock_connector,
    mock_ctx,
    runner,
    project_directory,
    project_name,
):
    mock_object_manager.return_value.show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
        {"name": "external_2", "type": "EXTERNAL_ACCESS"},
    ]
    mock_object_manager.return_value.describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_directory(project_name) as project_dir:
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
            catch_exceptions=False,
        )

    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string, b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=()
            external_access_integrations=(external_1, external_2)
            secrets=('cred'=cred_name, 'other'=other_name)
            """
        ).strip(),
    ]


@pytest.mark.parametrize(
    "project_name",
    [
        "snowpark_function_secrets_without_external_access",
        "snowpark_function_secrets_without_external_access_v2",
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function_secrets_without_external_access(
    mock_object_manager,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    os_agnostic_snapshot,
    project_name,
    enable_snowpark_glob_support_feature_flag,
):
    mock_object_manager.return_value.show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
        {"name": "external_2", "type": "EXTERNAL_ACCESS"},
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory(project_name):
        result = runner.invoke(
            [
                "snowpark",
                "build",
                "--ignore-anaconda",
            ],
            catch_exceptions=False,
        )
        assert result.exit_code == 0, result.output
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
        )

    assert result.exit_code == 1, result.output
    assert result.output == os_agnostic_snapshot


@pytest.mark.parametrize(
    "project_name", ["snowpark_functions", "snowpark_functions_v2"]
)
@mock.patch("snowflake.connector.connect")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_no_changes(
    mock_download,
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
    project_name,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    rows = [
        ("packages", '["foo==1.2.3", "bar>=3.0.0"]'),
        ("handler", "app.func1_handler"),
        ("returns", "string"),
        ("imports", "dev_deployment/my_snowpark_project/app.py"),
        ("runtime_version", "3.10"),
    ]

    queries, result, project_dir = _deploy_function(
        rows,
        mock_connector,
        runner,
        mock_ctx,
        mock_cursor,
        project_directory,
        project_name,
        "--replace",
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.func1(a string default 'default value', b variant)",
            "status": "packages updated",
            "type": "function",
        }
    ]
    assert queries == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        _put_query(
            Path(project_dir),
            "my_snowpark_project/app.py",
            "@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/",
        ),
    ]


@pytest.mark.parametrize(
    "project_name", ["snowpark_functions", "snowpark_functions_v2"]
)
@mock.patch("snowflake.connector.connect")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_needs_update_because_packages_changes(
    mock_download,
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
    project_name,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    rows = [
        ("packages", '["foo==1.2.3"]'),
        ("handler", "main.py:app"),
        ("returns", "table(variant)"),
    ]

    queries, result, project_dir = _deploy_function(
        rows,
        mock_connector,
        runner,
        mock_ctx,
        mock_cursor,
        project_directory,
        project_name,
        "--replace",
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.func1(a string default 'default value', b variant)",
            "status": "definition updated",
            "type": "function",
        }
    ]
    assert queries == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        _put_query(
            Path(project_dir),
            "my_snowpark_project/app.py",
            "@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/",
        ),
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string default 'default value', b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=('foo==1.2.3','bar>=3.0.0')
            """
        ).strip(),
    ]


@pytest.mark.parametrize(
    "project_name", ["snowpark_functions", "snowpark_functions_v2"]
)
@mock.patch("snowflake.connector.connect")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_needs_update_because_handler_changes(
    mock_download,
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
    project_name,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    rows = [
        ("packages", '["foo==1.2.3", "bar>=3.0.0"]'),
        ("handler", "main.py:oldApp"),
        ("returns", "table(variant)"),
    ]

    queries, result, project_dir = _deploy_function(
        rows,
        mock_connector,
        runner,
        mock_ctx,
        mock_cursor,
        project_directory,
        project_name,
        "--replace",
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.func1(a string default 'default value', b variant)",
            "status": "definition updated",
            "type": "function",
        }
    ]
    assert queries == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        _put_query(
            Path(project_dir),
            "my_snowpark_project/app.py",
            "@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/",
        ),
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string default 'default value', b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=('foo==1.2.3','bar>=3.0.0')
            """
        ).strip(),
    ]


@pytest.mark.parametrize(
    "project_name",
    [
        "snowpark_function_fully_qualified_name",
        "snowpark_function_fully_qualified_name_v2",
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_fully_qualified_name_duplicated_database(
    mock_download,
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
    project_name,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    number_of_functions_in_project = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_functions_in_project
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory(project_name):
        result = runner.invoke(
            [
                "snowpark",
                "build",
                "--ignore-anaconda",
            ],
            catch_exceptions=False,
        )
        assert result.exit_code == 0, result.output
        result = runner.invoke(["snowpark", "deploy"])
        assert result.output == os_agnostic_snapshot(name="database error")


@pytest.mark.parametrize(
    "project_name,path_in_project_file",
    [
        ("snowpark_function_fully_qualified_name", "snowpark.functions.5.name"),
        (
            "snowpark_function_fully_qualified_name_v2",
            "entities.custom_database_custom_schema_fqn_function_error.identifier.name",
        ),
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_fully_qualified_name_duplicated_schema(
    mock_download,
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
    project_name,
    path_in_project_file,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    number_of_functions_in_project = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_functions_in_project
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory(project_name) as tmp_dir:
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path=path_in_project_file,
            value="custom_schema.fqn_function_error",
        )
        result = runner.invoke(
            [
                "snowpark",
                "build",
                "--ignore-anaconda",
            ],
            catch_exceptions=False,
        )
        assert result.exit_code == 0, result.output
        result = runner.invoke(["snowpark", "deploy"])
        assert result.output == os_agnostic_snapshot(name="schema error")


@pytest.mark.parametrize(
    "project_name,parameter_path",
    [
        ("snowpark_function_fully_qualified_name", "snowpark.functions.5.name"),
        (
            "snowpark_function_fully_qualified_name_v2",
            "entities.custom_database_custom_schema_fqn_function_error.identifier.name",
        ),
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_fully_qualified_name(
    mock_download,
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
    project_name,
    parameter_path,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    number_of_functions_in_project = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_functions_in_project
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory(project_name) as tmp_dir:
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path=parameter_path,
            value="fqn_function3",
        )
        result = runner.invoke(
            [
                "snowpark",
                "build",
                "--ignore-anaconda",
            ],
            catch_exceptions=False,
        )
        assert result.exit_code == 0, result.output
        result = runner.invoke(["snowpark", "deploy"])
        assert result.exit_code == 0
        assert result.output == os_agnostic_snapshot(name="ok")


@pytest.mark.parametrize(
    "project_name,signature_path,runtime_path",
    [
        (
            "snowpark_functions",
            "snowpark.functions.0.signature.0.",
            "snowpark.functions.0.runtime",
        ),
        (
            "snowpark_functions_v2",
            "entities.func1.signature.0.",
            "entities.func1.runtime",
        ),
    ],
)
@pytest.mark.parametrize(
    "parameter_type,default_value",
    [
        ("string", None),
        ("string", ""),
        ("int", None),
        ("variant", None),
        ("bool", None),
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock.patch(
    "snowflake.cli._plugins.snowpark.package_utils.download_unavailable_packages"
)
@mock_session_has_warehouse
def test_deploy_function_with_empty_default_value(
    mock_download,
    mock_object_manager,
    mock_connector,
    mock_ctx,
    runner,
    project_directory,
    alter_snowflake_yml,
    parameter_type,
    default_value,
    project_name,
    signature_path,
    runtime_path,
    enable_snowpark_glob_support_feature_flag,
):
    mock_download.return_value = DownloadUnavailablePackagesResult()
    mock_object_manager.return_value.describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    with project_directory(project_name) as project_dir:
        snowflake_yml = project_dir / "snowflake.yml"
        for param, value in [("type", parameter_type), ("default", default_value)]:
            alter_snowflake_yml(
                snowflake_yml,
                parameter_path=f"{signature_path}{param}",
                value=value,
            )
        alter_snowflake_yml(
            snowflake_yml,
            parameter_path=runtime_path,
            value="3.10",
        )
        result = runner.invoke(
            [
                "snowpark",
                "build",
                "--ignore-anaconda",
            ],
            catch_exceptions=False,
        )
        assert result.exit_code == 0, result.output
        result = runner.invoke(
            ["snowpark", "deploy", "--format", "json"], catch_exceptions=False
        )
    default_value_json = default_value
    if default_value is None:
        default_value_json = "null"
    elif parameter_type == "string":
        default_value_json = f"'{default_value}'"

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": f"MockDatabase.MockSchema.func1(a {parameter_type} default {default_value_json}, b variant)",
            "status": "created",
            "type": "function",
        }
    ]


@mock.patch("snowflake.connector.connect")
def test_execute_function(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    result = runner.invoke(
        [
            "snowpark",
            "execute",
            "function",
            "functionName(42, 'string')",
        ]
    )

    assert result.exit_code == 0, result.output
    assert ctx.get_query() == "select functionName(42, 'string')"


def _deploy_function(
    rows,
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
    project_name,
    *args,
):
    ctx = mock_ctx(mock_cursor(rows=rows, columns=[]))
    mock_connector.return_value = ctx
    with (
        mock.patch(
            "snowflake.cli._plugins.snowpark.commands.ObjectManager.describe"
        ) as om_describe,
        mock.patch(
            "snowflake.cli._plugins.snowpark.commands.ObjectManager.show"
        ) as om_show,
    ):
        om_describe.return_value = rows

        with project_directory(project_name) as temp_dir:
            (Path(temp_dir) / "requirements.snowflake.txt").write_text(
                "foo==1.2.3\nbar>=3.0.0"
            )
            result = runner.invoke(
                [
                    "snowpark",
                    "build",
                    "--ignore-anaconda",
                ],
                catch_exceptions=False,
            )
            assert result.exit_code == 0, result.output
            result = runner.invoke(
                [
                    "snowpark",
                    "deploy",
                    "--format",
                    "json",
                    *args,
                ]
            )
    queries = ctx.get_queries()
    return queries, result, temp_dir


@mock.patch("snowflake.connector.connect")
@pytest.mark.parametrize(
    "command, parameters",
    [
        ("list", []),
        ("list", ["--like", "PATTERN"]),
        ("describe", ["NAME"]),
        ("drop", ["NAME"]),
    ],
)
def test_command_aliases(mock_connector, runner, mock_ctx, command, parameters):
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    result = runner.invoke(["object", command, "function", *parameters])
    assert result.exit_code == 0, result.output
    result = runner.invoke(
        ["snowpark", command, "function", *parameters], catch_exceptions=False
    )
    assert result.exit_code == 0, result.output

    queries = ctx.get_queries()
    assert queries[0] == queries[1]


def _put_query(project_root: Path, source: str, dest: str):
    return dedent(
        f"put file://{project_root.resolve() / 'output' / 'bundle' / 'snowpark' / source} {dest} auto_compress=false parallel=4 overwrite=True"
    )
