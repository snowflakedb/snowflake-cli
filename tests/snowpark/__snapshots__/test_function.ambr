# serializer version: 1
# name: test_deploy_function_fully_qualified_name[snowpark_function_fully_qualified_name-snowpark.functions.5.name][ok]
  '''
  Performing initial validation
  Checking remote state
  Preparing required stages and artifacts
    Creating (if not exists) stage: dev_deployment
    Uploading app.zip to @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/
  Creating Snowpark entities
    Creating function custom_db.custom_schema.fqn_function
    Creating function custom_schema.fqn_function_only_schema
    Creating function custom_schema.schema_function
    Creating function custom_db.PUBLIC.database_function
    Creating function custom_db.custom_schema.database_function
    Creating function custom_database.custom_schema.fqn_function3
  +------------------------------------------------------------------------------+
  | object                                                  | type     | status  |
  |---------------------------------------------------------+----------+---------|
  | custom_db.custom_schema.fqn_function(name string)       | function | created |
  | MockDatabase.custom_schema.fqn_function_only_schema(nam | function | created |
  | e string)                                               |          |         |
  | MockDatabase.custom_schema.schema_function(name string) | function | created |
  | custom_db.MockSchema.database_function(name string)     | function | created |
  | custom_db.custom_schema.database_function(name string)  | function | created |
  | custom_database.custom_schema.fqn_function3(name        | function | created |
  | string)                                                 |          |         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name[snowpark_function_fully_qualified_name_v2-entities.custom_database_custom_schema_fqn_function_error.identifier.name][ok]
  '''
  Performing initial validation
  Checking remote state
  Preparing required stages and artifacts
    Creating (if not exists) stage: dev_deployment
    Uploading app.zip to @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/
  Creating Snowpark entities
    Creating function custom_database.custom_schema.fqn_function3
    Creating function custom_db.custom_schema.fqn_function
    Creating function custom_db.custom_schema.database_function
    Creating function custom_schema.fqn_function_only_schema
    Creating function custom_db.PUBLIC.database_function
    Creating function custom_schema.schema_function
  +------------------------------------------------------------------------------+
  | object                                                  | type     | status  |
  |---------------------------------------------------------+----------+---------|
  | custom_database.custom_schema.fqn_function3(name        | function | created |
  | string)                                                 |          |         |
  | custom_db.custom_schema.fqn_function(name string)       | function | created |
  | custom_db.custom_schema.database_function(name string)  | function | created |
  | MockDatabase.custom_schema.fqn_function_only_schema(nam | function | created |
  | e string)                                               |          |         |
  | custom_db.MockSchema.database_function(name string)     | function | created |
  | MockDatabase.custom_schema.schema_function(name string) | function | created |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name_duplicated_database[snowpark_function_fully_qualified_name][database error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Database provided but name                                                   |
  | 'custom_database.custom_schema.fqn_function_error' is fully qualified name.  |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name_duplicated_database[snowpark_function_fully_qualified_name_v2][database error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Database provided but name                                                   |
  | 'custom_database.custom_schema.fqn_function_error' is fully qualified name.  |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name_duplicated_schema[snowpark_function_fully_qualified_name-snowpark.functions.5.name][schema error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Schema provided but name 'custom_schema.fqn_function_error' is fully         |
  | qualified name.                                                              |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name_duplicated_schema[snowpark_function_fully_qualified_name_v2-entities.custom_database_custom_schema_fqn_function_error.identifier.name][schema error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Schema provided but name 'custom_schema.fqn_function_error' is fully         |
  | qualified name.                                                              |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_secrets_without_external_access[snowpark_function_secrets_without_external_access]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | func1 defined with secrets but without external integration.                 |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_secrets_without_external_access[snowpark_function_secrets_without_external_access_v2]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | func1 defined with secrets but without external integration.                 |
  +------------------------------------------------------------------------------+
  
  '''
# ---
